### PR TITLE
Add Pulp Perf Test Case

### PR DESCRIPTION
**Pull request by danuzclaudes**

### CODE REVIEW

File: robottelo.properties.sample [hunk 1]
```diff
@@ -127,3 +127,13 @@ test.virtual_machines_list=127.0.0.1,127.0.0.1,127.0.0.1,127.0.0.1,127.0.0.1
 # Class `ConcurrentTestCase` and its subclasses use this setting when
 # computing statistics of each performance test case, grouped in buckets.
 csv.num_buckets=10
+
+# Target repository names to be synchronized by Pulp.
+test.target_repos=Repo Example 1,Repo-Example-2
```
Comment (Ichimonji10) [inline robottelo.properties.sample]: Can you do one of these?

    test.target_repos=
    # test.target_repos=

If the user must provide a value for some field and the user fails to provide a value, then we want an error to be thrown as soon as reasonably possible. Providing a bogus default value runs counter to that goal.

In other words:

* If a value must be provided and we can give a sane default, give a sane default.
* If a value must be provided and we cannot give a sane deafult, don't give an insane default.

File: robottelo.properties.sample [hunk 2]
```diff
@@ -127,3 +127,13 @@ test.virtual_machines_list=127.0.0.1,127.0.0.1,127.0.0.1,127.0.0.1,127.0.0.1
...
+
+# Number of times to repeat synchronization on each repository
+test.num_syncs=3
+
+# Parameter for deciding whether conduct initial sync or resync
+# 0 denotes resync; 1 denotes initial sync
+test.is_initial_sync=0
```
Comment (Ichimonji10) [inline robottelo.properties.sample]: You're not limited to integers. Why not provide a string?

    test.test_type='resync'
    test.test_type='sync'

File: robottelo/performance/candlepin.py [hunk 3]
```diff
@@ -29,7 +29,7 @@ class Candlepin(object):
     serverhost = conf.properties['main.server.hostname']
 
     @classmethod
-    def get_real_time(result):
+    def get_real_time(cls, result):
```
Comment (Ichimonji10) [inline robottelo/performance/candlepin.py]: You're not using the `cls` argument in this method. So why bother making this a classmethod? How about making this a staticmethod?

File: robottelo/performance/graph.py [hunk 4]
```diff
@@ -93,15 +47,15 @@ def generate_stacked_line_chart_raw(time_result_dict, head, filename):
     stackedline_chart.render_to_file(filename)
 
 
-def generate_line_chart_raw(time_result_dict, head, filename):
-    """Generate Normal Line chart for raw data of ak/att/del/reg
+def generate_line_chart_raw(time_result_dict, head, filename, line_chart):
+    """Common function for line charts used by Candlepin and Pulp tests
```
Comment (Ichimonji10) [inline robottelo/performance/graph.py]: This docstring doesn't tell me what the function does. Don't tell me that this function contains common functionality - of course it does. It's very common for a function to do that. Do tell me what this function does.

File: robottelo/performance/graph.py [hunk 5]
```diff
@@ -113,3 +67,101 @@ def generate_line_chart_raw(time_result_dict, head, filename):
         time_list = time_result_dict.get(key)
         line_chart.add('client-{}'.format(thread), time_list)
     line_chart.render_to_file(filename)
+
+
+def generate_line_chart_raw_Pulp(time_result_dict, head, filename):
```
Comment (Ichimonji10) [inline robottelo/performance/graph.py]: Use lower case method names.

File: robottelo/performance/graph.py [hunk 6]
```diff
@@ -113,3 +67,101 @@ def generate_line_chart_raw(time_result_dict, head, filename):
...
+        line_chart):
+    """Common funtion for Line charts used by Candlepin and Pulp tests
+
+    Take a dictionary of statistic values on each client. For example::
+
+    0: [min, | median,max,std]
+    1: [min, | median,max,std]
+    2: [min, | median,max,std]
+    ...
+    9: [min, | median,max,std],
```
Comment (Ichimonji10) [inline robottelo/performance/graph.py]: Indent code blocks.

File: robottelo/performance/graph.py [hunk 7]
```diff
@@ -113,3 +67,101 @@ def generate_line_chart_raw(time_result_dict, head, filename):
...
+def generate_line_chart_raw_Candlepin(time_result_dict, head, filename):
+    """Generate Normal Line chart for raw data of ak/att/del/reg"""
+    line_chart = pygal.Line(show_dots=False, range=(0, 50))
+    generate_line_chart_raw(time_result_dict, head, filename, line_chart)
+
+
+def generate_line_chart_stat(
+        stat_dict,
+        filename,
+        line_chart):
```
Comment (Ichimonji10) [inline robottelo/performance/graph.py]: This entier function signature fits on one line.

File: robottelo/performance/graph.py [hunk 8]
```diff
@@ -113,3 +67,101 @@ def generate_line_chart_raw(time_result_dict, head, filename):
...
     line_chart.render_to_file(filename)
+
+
+def generate_line_chart_raw_Pulp(time_result_dict, head, filename):
+    """Generate Normal Line chart for raw data of sync/resync"""
+    line_chart = pygal.Line()
+    generate_line_chart_raw(time_result_dict, head, filename, line_chart)
+
+
+def generate_line_chart_raw_Candlepin(time_result_dict, head, filename):
```
Comment (Ichimonji10) [inline robottelo/performance/graph.py]: User lower case method names.

File: robottelo/performance/pulp.py [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+            )
+            return 0
+        LOGGER.info(
+            'Sync repository {0} by thread-{1} successful!'
+            .format(repo_name, thread_id)
+        )
+        return cls.get_elapsed_time(result.stderr)
+
+    @classmethod
+    def get_elapsed_time(cls, stderr):
```
Comment (Ichimonji10) [inline robottelo/performance/pulp.py]: This method uses nothing on the class. Why bother making it a class method? How about making this a static method instead?

File: robottelo/performance/pulp.py [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+        """retrieve time from stderr"""
+        # should return only one time point as a single sync
+        real_time = ''
+        for line in stderr.split('\n'):
+            if line.startswith('real'):
+                real_time = line
+        return 0 if real_time == '' else float(real_time.split(' ')[1])
+
+    @classmethod
+    def _get_enabled_repos(cls, org_id):
```
Comment (Ichimonji10) [inline robottelo/performance/pulp.py]: This method uses nothing on the class. Why bother making it a class method? How about making this a static method instead?

File: robottelo/performance/pulp.py [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+            if savepoint is None:
+                return
+            else:
+                # restore database at the end of each iteration
+                cls._restore_from_savepoint(savepoint)
+
+        return time_result_dict_sync
+
+    @classmethod
+    def _restore_from_savepoint(cls, savepoint):
```
Comment (Ichimonji10) [inline robottelo/performance/pulp.py]: This method uses nothing on the class. Why bother making it a class method? How about making this a static method instead?

**Comment by Ichimonji10** [comment]
Please run pylint over the modules you've changed and update this thread when you've done so. I'll continue reviewing then.

File: robottelo.properties.sample (hunk 1 above)
Comment (danuzclaudes) [inline robottelo.properties.sample]: @Ichimonji10, I actually read this entry here: https://github.com/danuzclaudes/robottelo/commit/b97b2635be1d1caf7105657664dd9b57bd582a72#diff-30051cc3cf62fb3c2573508e5d08d99cR88
But as you've said, I'd raise exception if the field is empty. BTW, if the user provides invalid repository name, the synchronization would just ignore it and continue next.

File: robottelo/performance/candlepin.py (hunk 3 above)
Comment (danuzclaudes) [inline robottelo/performance/candlepin.py]: @Ichimonji10, in fact, this is for bug-fixing. I previously missed the cls there.

File: robottelo/performance/graph.py (hunk 4 above)
Comment (danuzclaudes) [inline robottelo/performance/graph.py]: @Ichimonji10, ok i see it

File: robottelo/performance/pulp.py (hunk 10 above)
Comment (danuzclaudes) [inline robottelo/performance/pulp.py]: @Ichimonji10, this method is invoked by https://github.com/danuzclaudes/robottelo/commit/b97b2635be1d1caf7105657664dd9b57bd582a72#diff-30051cc3cf62fb3c2573508e5d08d99cR56, as a utility function of Pulp class.

File: robottelo/performance/pulp.py (hunk 11 above)
Comment (danuzclaudes) [inline robottelo/performance/pulp.py]: So I extract this out from https://github.com/danuzclaudes/robottelo/blob/perf-pulp/robottelo/performance/pulp.py#L129, i.e. above it. But I can make it static if you want.

File: robottelo/performance/pulp.py (hunk 9 above)
Comment (danuzclaudes) [inline robottelo/performance/pulp.py]: OK, I can make it static.

File: robottelo/performance/graph.py [hunk 12]
```diff
@@ -93,15 +47,15 @@ def generate_stacked_line_chart_raw(time_result_dict, head, filename):
...
 
     :param dict stat_dict: The dictionary containing min/median/max/std
     :param str head: Titile of charts
     :param str filename: The name of output svg chart
+    :param obj line_chart: Line chart object from either Pulp or Candlepin
 
     """
-    line_chart = pygal.Line(show_dots=False, range=(0, 50))
     max_label = len(time_result_dict.get('thread-0'))
     line_chart.x_labels = [str(i) for i in range(1, max_label + 1)]
```
Comment (rplevka) [inline robottelo/performance/graph.py]: :+1:

File: robottelo.properties.sample (hunk 1 above)
Comment (Ichimonji10) [inline robottelo.properties.sample]: > @Ichimonji10, I actually read this entry here: https://github.com/danuzclaudes/robottelo/commit/b97b2635be1d1caf7105657664dd9b57bd582a72#diff-30051cc3cf62fb3c2573508e5d08d99cR88

Understood. But that's irrelevant.

> But as you've said, I'd raise exception if the field is empty.

Sure, you can check to see if `test.target_repos == ''`. That's a reasonable level of user-friendliness.

> BTW, if the user provides invalid repository name, the synchronization would just ignore it and continue next.

Again, providing a bogus value runs counter to the goal of letting the user know about their responsibilities (i.e. filling in the config file) as soon as possible.

File: robottelo/performance/candlepin.py (hunk 3 above)
Comment (Ichimonji10) [inline robottelo/performance/candlepin.py]: You can also fix this bug by just adding `@staticmethod`.

File: robottelo/performance/pulp.py (hunk 9 above)
Comment (Ichimonji10) [inline robottelo/performance/pulp.py]: > OK. It's actually invoked by line 51 above, but I can make it static.

*sigh*

**This method** uses nothing on the class. This method does not use the `cls` variable.

File: robottelo/performance/pulp.py (hunk 10 above)
Comment (Ichimonji10) [inline robottelo/performance/pulp.py]: You're completely missing my point. Read about [staticmethod()](https://docs.python.org/3.4/library/functions.html#staticmethod). Read about [classmethod()](https://docs.python.org/3.4/library/functions.html#classmethod). Google around a bit and learn the difference between:

* instance methods
* class methods
* static methods

And then re-read my comment.

File: robottelo/performance/pulp.py (hunk 11 above)
Comment (Ichimonji10) [inline robottelo/performance/pulp.py]: That's not what I'm asking for. I'm pointing out that the `cls` argument is unused, so it should be dropped and the `@staticmethod` decorator should be applied.

Learn about:

* instance method
* class methods
* instance methods

Then come back and re-read my comment.

File: robottelo/performance/graph.py [hunk 13]
```diff
@@ -113,3 +67,101 @@ def generate_line_chart_raw(time_result_dict, head, filename):
...
+
+def generate_line_chart_raw_Pulp(time_result_dict, head, filename):
+    """Generate Normal Line chart for raw data of sync/resync"""
+    line_chart = pygal.Line()
+    generate_line_chart_raw(time_result_dict, head, filename, line_chart)
+
+
+def generate_line_chart_raw_Candlepin(time_result_dict, head, filename):
+    """Generate Normal Line chart for raw data of ak/att/del/reg"""
+    line_chart = pygal.Line(show_dots=False, range=(0, 50))
```
Comment (rplevka) [inline robottelo/performance/graph.py]: fixed range again :(
could you calculate the `max_label` as you did at line [#80](https://github.com/SatelliteQE/robottelo/pull/2657/files#diff-0065083de42132eb44a84463371d1320R80)?

**Comment by danuzclaudes** [comment]
@Ichimonji10, I have updated all your previous comments and linted the performance modules. Thanks.

**Comment by danuzclaudes** [comment]
@rplevka, as for the range of Line chart, I explained here: https://github.com/SatelliteQE/robottelo/pull/2657/files#diff-0065083de42132eb44a84463371d1320R54, FYI. Thanks!

File: robottelo/performance/pulp.py [hunk 14]
```diff
@@ -0,0 +1,140 @@
...
+        for i in range(sync_iterations):
+            # note: name key by thread to adapt to graph module
+            key = 'thread-{}'.format(i)
+            time_result_dict_sync[key] = []
+
+            # Sync each repo one-by-one and collect timing data
+            for repo_name in repo_names_list:
+                repo_id = map_repo_name_id.get(repo_name, None)
+                if repo_id is None:
+                    LOGGER.warning('Invalid repository name!')
```
Comment (Ichimonji10) [inline robottelo/performance/pulp.py]: Include `repo_name` in the warning message.

File: robottelo/performance/graph.py [hunk 15]
```diff
@@ -113,3 +78,94 @@ def generate_line_chart_raw(time_result_dict, head, filename):
...
+    :param dict stat_dict: The dictionary containing min/median/max/std
+    :param str head: Titile of charts
+    :param str filename: The name of output svg chart
+    :param obj line_chart: Line chart object from either Pulp or Candlepin
+
+    """
+    # parameters for formats on charts
+    lengend = ('min', 'median', 'max', 'std')
+    for i in range(len(lengend)):
+        line_chart.add(lengend[i], [vlist[i] for vlist in stat_dict.values()])
```
Comment (Ichimonji10) [inline robottelo/performance/graph.py]: Nitpick: `s/lengend/legend/`

File: robottelo/performance/pulp.py [hunk 16]
```diff
@@ -0,0 +1,140 @@
...
+        """
+        LOGGER.info('Get All enabled repositories from database:')
+
+        result = Repository.list(
+            {'organization-id': org_id},
+            per_page=False
+        )
+
+        if result.return_code != 0:
+            LOGGER.info('Fail to fetch enabled repos.')
```
Comment (Ichimonji10) [inline robottelo/performance/pulp.py]: IMO, it doesn't make sense to log an info-priority message right before raising a `RuntimeError`. If the exception is uncaught, then the exception message will make its way in to the logs. And if the caller catches the exception, then there's little reason to log anything out - the caller addressed the situation.

File: robottelo/performance/pulp.py [hunk 17]
```diff
@@ -0,0 +1,140 @@
...
+            raise RuntimeError('No enabled repository found!')
```
Comment (Ichimonji10) [inline robottelo/performance/pulp.py]: Can you include information about the search in this message? I see above that you're searching for repositories in a particular organization. How about including that information in this message?

File: robottelo/performance/pulp.py [hunk 18]
```diff
@@ -0,0 +1,140 @@
...
+                real_time = line
+        return 0 if real_time == '' else float(real_time.split(' ')[1])
+
+    @staticmethod
+    def get_enabled_repos(org_id):
+        """Get all enabled repositories ids and names
+
+        :return map_repo_name_id: The dictionary contains all enabled
+            repositories in Satellite. Map repo-name as key, repo-id as value
+
```
Comment (Ichimonji10) [inline robottelo/performance/pulp.py]: You can document exceptions in docstrings. For example:

    :raises: ``RuntimeException`` if the knob is excessively frobnicated.

Or, if you've defined a custom exception that Sphinx can reference:

    :raises robottelo.performance.pulp.MyCustomException: If the knob is excessively frobnicated.

File: robottelo/performance/pulp.py [hunk 19]
```diff
@@ -0,0 +1,140 @@
...
+        """
+        LOGGER.info('Get All enabled repositories from database:')
```
Comment (Ichimonji10) [inline robottelo/performance/pulp.py]: I think the user reading these log messages will expect to see messages like this scrolling by:

> Searching for enabled repositories.

Or maybe:

> Fetching all enabled repositories from the database.

Can you fiddle with the wording here a little bit? Maybe include information about what sort of search is being done?

File: tests/foreman/performance/test_pulp_sync.py [hunk 20]
```diff
@@ -0,0 +1,281 @@
...
+)
+from robottelo.performance.pulp import Pulp
+from robottelo.performance.stat import generate_stat_for_pulp_sync
+from robottelo.test import ConcurrentTestCase
+
+
+class ConcurrentSyncTestCase(ConcurrentTestCase):
+    """Concurrent Sync with Red Hat CDN
+
+    Initial-Sync test would restore databse between each
```
Comment (Ichimonji10) [inline tests/foreman/performance/test_pulp_sync.py]: Nitpick: `s/databse/database/`

File: tests/foreman/performance/test_pulp_sync.py [hunk 21]
```diff
@@ -0,0 +1,281 @@
...
+        )
+        cls.is_initial_sync = True if sync_parameter == 'sync' else False
+
+    def setUp(self):
+        super(ConcurrentSyncTestCase, self).setUp()
+
+        # determine all targeting repositories to be synced
+        self.repo_names_list = self._get_target_repo_from_config()
+        self.logger.debug(
+            "Target Repositories to be synced: {}"
```
Comment (Ichimonji10) [inline tests/foreman/performance/test_pulp_sync.py]: Nitpick: use single quotes.

File: tests/foreman/performance/test_pulp_sync.py [hunk 22]
```diff
@@ -0,0 +1,281 @@
...
+            if repo in self.map_repo_name_id
+        ]
+        return repo_names_list
+
+    def test_concurrent_synchronization(self):
+        """
+        @Test: Synchronize two repos concurrently
+
+        @Steps:
+        1. get list of all enabled repositories (setUpClass)
```
Comment (Ichimonji10) [inline tests/foreman/performance/test_pulp_sync.py]: Lists should be separated from other paragraphs with a blank line. Just add one line above the `1. get list of all ...` like so:

    @Steps:

    1. get list of all enabled repositories (setUpClass)

File: tests/foreman/performance/test_pulp_sync.py [hunk 23]
```diff
@@ -0,0 +1,281 @@
...
+            'perf-statistics-concurrent-sync.svg',
+            self.max_num_tests
+        )
+
+    def test_sequential_synchronization(self):
+        """
+        @Test: Synchronize two repos sequentially
+
+        @Steps:
+        1. get list of all enabled repositories (setUpClass)
```
Comment (Ichimonji10) [inline tests/foreman/performance/test_pulp_sync.py]: Add a blank line between `@Steps:` and the beginning of the list, like so:

    @Steps:

    1. get list of all enabled repositories (setUpClass)

File: robottelo/test.py [hunk 24]
```diff
@@ -991,3 +987,86 @@ def kick_off_del_test(self, current_num_threads):
...
+
+                if repo_id is None:
+                    self.logger.warning('Invalid repository name!')
+                    continue
+
+                self.logger.debug(
+                    '{0} repository {1} attempt {2} '
+                    'on {3}-repo test case starts:'
+                    .format(
+                        'Initially sync' if is_initial_sync else 'Resync',
```
Comment (Ichimonji10) [inline robottelo/test.py]: Should this be changed to just `is_initial_sync`?

File: robottelo.properties.sample [hunk 25]
```diff
@@ -127,3 +127,16 @@ test.virtual_machines_list=127.0.0.1,127.0.0.1,127.0.0.1,127.0.0.1,127.0.0.1
...
+# Real repository names should be referred by
+# `h repository list --organization-id=1`
+test.target_repos=RHEL 6 Kickstart rpms,RHEL 6 Optional Source rpms
+
+# Number of times to repeat synchronization on each repository
+test.num_syncs=3
+
+# Parameter for deciding whether conduct initial sync or resync
+# 'resync' denotes resync; 'sync' denotes initial sync
+test.is_initial_sync='sync'
```
Comment (Ichimonji10) [inline robottelo.properties.sample]: Would it make even more sense to use a name like `sync_type`?

    test.sync_type='sync'

**Comment by Ichimonji10** [comment]
Nice improvement, @danuzclaudes. ACK pending comments.

File: robottelo/test.py (hunk 24 above)
Comment (danuzclaudes) [inline robottelo/test.py]: @Ichimonji10, the `is_initial_sync` is set by: https://github.com/danuzclaudes/robottelo/blob/perf-pulp/tests/foreman/performance/test_pulp_sync.py#L70, therefore a boolean.

Comment (Ichimonji10) [inline robottelo/test.py]: :+1:

**Comment by danuzclaudes** [comment]
@Ichimonji10, all comments reflected. Also got rebased to upstream. Thank you!

**Comment by Ichimonji10** [comment]
Thank you. ACK

**Comment by elyezer** [comment]
ACK, good stuff.

**Pull request merged**